lib/edgehog_device_forwarder/tokens.ex

```diff
@@ -0,0 +1,66 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.Tokens do
+  alias EdgehogDeviceForwarder.Tokens.Data
+  alias EdgehogDeviceForwarder.TerminationCallbacks
+  require Logger
+
+  @tokens EdgehogDeviceForwarder.Cache.for(__MODULE__)
+  @type token :: String.t()
+  @type device_id :: Data.device_id()
+  @type realm :: Data.realm()
+  @type device_socket :: Data.device_socket()
+
+  @spec add_device_info(token, device_id, realm) ::
+          :ok | {:error, :token_not_found}
+  def add_device_info(token, device_id, realm) do
+    ConCache.update(@tokens, token, fn
+      nil -> {:error, :token_not_found}
+      token_data -> {:ok, Data.put_device_info(token_data, device_id, realm)}
+    end)
+  end
+
+  @spec reserve(token) :: :ok | {:error, :token_already_exists}
+  def reserve(token) do
+    with {:error, :already_exists} <- ConCache.insert_new(@tokens, token, :reserved) do
+      {:error, :token_already_exists}
+    end
```

Comment:
I would argue that something like this may be more readable: you explicitly show the happy path (given that `reserve` is not meant to be used only as a step in a `with` block):
```suggestion
    case ConCache.insert_new(@tokens, token, :reserved) do
      :ok -> :ok
      {:error, :already_exists} -> {:error, :token_already_exists}
    end
```

Reply:
Done